packages/component-aws-s3/README.md

```diff
@@ -0,0 +1,227 @@
+    s3Callbacks?: {
+        /* Called when a new active file is added to S3 */
+        newActiveFile: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+
+        /* Called when a new pending (still undergoing workflow processing) file is added to S3 */
+        newPendingFile: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+
+        /* Called when pending file becomes active */
+        fileActivated: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+
+        /* Called when a file is deleted from S3 */
+        fileDeleted: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+    },
+    //... remainder of S3 Config
+}
+```
+
+### Workflow to Upload a String to S3
+The `saveStringToFile` workflow stores a string to an S3 key, and runs the callback function to update the database.  If anything goes wrong during the workflow, S3 will be cleaned up and the database will be unchanged by the workflow.
+
+```typescript
+    await ctx.invokeWorkflow(defaultS3).saveStringToFile(fileDBRecord, 'This is my file');
+```
+
+This workflow performs the following actions:
+* Puts the string in S3
+* If there is difficulty with S3, ensures that no entry is left there and throws an error
+* Invokes the callback for a new active file record
+
+***Note:*** **Function arguments to workflows, transactions, and communicators are logged to the database.  This means that workflow functions are suitable for writing kilobytes of data to S3, that megabytes will possibly work, but gigabytes will cause great pain.  Consider [having the client upload to S3 with a presigned post](#workflow-to-allow-client-file-upload) if the data is large.**
+
+### Workflow to Retrieve a String from S3
+The workflow function `readStringFromFile(ctx: WorkflowContext, fileDetails: FileRecord)` will retrieve the contents of an S3 object as a `string`.
+
+This workflow currently performs no additional operations outside of a call to `readStringFromFileComm(fileDetails.key)`.
+
+***Note:*** **Function return values from workflows, transactions, and communicators are logged to the database.  This means that workflow functions are suitable for reading kilobytes of data from S3, that megabytes will possibly work, but gigabytes will cause great pain.  Consider [having the client read from S3 with a signed URL](#workflow-to-allow-client-file-download) if the data is large.**
+
+### Workflow to Delete a File
+The `deleteFile` workflow function removes a file from both S3 and the database.
+
+```typescript
+await ctx.invokeWorkflow(defaultS3).deleteFile(fileDBRecord);
+```
+
+This workflow performs the following actions:
+* Invokes the callback for a deleted file record
+* Removes the key from S3
+
+### Workflow to Allow Client File Upload
+The workflow to allow application clients to upload to S3 directly is more involved than other workflows, as it runs concurrently with the client.  The workflow interaction generally proceeds as follows:
```

Comment:
This needs to be radically simplified. Is it possible to avoid direct calls to the messaging API?  For example by wrapping them in other library methods?